cmd/containerd-shim-runhcs-v1/task_hcs.go

```diff
@@ -1008,3 +1026,72 @@ func (ht *hcsTask) requestUpdateContainer(ctx context.Context, resourcePath stri
 	}
 	return ht.c.Modify(ctx, modification)
 }
+
+// publishNotifications monitors the containers notification queue.
+// Currently only OOM events are published.
+//
+// This MUST be called via a goroutine to wait on a background thread.
+func (ht *hcsTask) publishNotifications(ctx context.Context) {
+	var err error
+	ctx, span := trace.StartSpan(ctx, "hcsTask::publishNotifications")
+	defer span.End()
+	defer func() { oc.SetSpanStatus(span, err) }()
+	span.AddAttributes(trace.StringAttribute(logfields.TaskID, ht.id))
+	entity := log.G(ctx).WithField(logfields.TaskID, ht.id)
+
+	nq, err := ht.c.Notifications()
+	if err == notifications.ErrNotificationsNotSupported {
+		entity.Debug("container does not support notifications")
+		return
+	} else if err != nil {
+		entity.WithError(err).Warning("could not access notification queue")
+		return
+	}
+
+outter:
+	for {
+		select {
+		case <-ctx.Done():
+			break outter
+		default:
+			v, err := nq.ReadOrWait()
+			if err != nil {
+				if err == queue.ErrQueueClosed {
+					// set err to nil for oc.SetSpanStatus
+					err = nil //nolint:ineffassign
```

Comment:
I believe this `err` is different from the one at the top of the function that the SetSpanStatus calls because you use a `:=` to assign the value from `ReadOrWait`, so this wouldn't do anything and in the case of an actual error in here, that would not be added to the span status on exit of the function.

Reply:
good catch, thanks!